drivers/clock_control/clock_control_rcar_cpg_mssr.c

```diff
@@ -0,0 +1,226 @@
+				    uint32_t module, uint32_t rate, bool enable)
+{
+	const struct rcar_mssr_config *config = DEV_CFG(dev);
+	uint32_t addr, divider;
+	int key, ret;
+
+	/* Only support CANFD core clock at the moment */
+	if (module != CPG_CORE_CLK_CANFD) {
+		return -EINVAL;
+	}
+
+	key = irq_lock();
+
+	addr = config->base_address + CANFDCKCR;
+	if (enable) {
+		if (CANFDCKCR_PARENT_CLK_RATE % rate) {
+			__ASSERT(true, "Can not generate "
+				 "%u from CANFD parent clock", rate);
+			ret = -EINVAL;
+			goto unlock;
+		}
+		divider = (CANFDCKCR_PARENT_CLK_RATE / rate) - 1;
+		if (divider > CANFDCKCR_DIVIDER_MASK) {
+			__ASSERT(true, "Can not generate %u "
+				 "from CANFD parent clock", rate);
+			ret = -EINVAL;
+			goto unlock;
+		}
+		cpg_write(addr, divider);
+	} else {
+		cpg_write(addr, CANFDCKCR_CKSTP);
+	}
+
+unlock:
+	irq_unlock(key);
+	return 0;
+}
+
+static int cpg_rmstp_clock_endisable(const struct device *dev,
+				     uint32_t module, bool enable)
+{
+	const struct rcar_mssr_config *config = DEV_CFG(dev);
+	uint32_t reg = module / 100;
+	uint32_t bit = module % 100;
+	uint32_t bitmask = BIT(bit);
+	uint32_t reg_val;
+
+	int key = irq_lock();
+
+	reg_val = sys_read32(config->base_address + rmstpsr[reg]);
+	if (enable) {
+		reg_val &= ~bitmask;
+	} else {
+		reg_val |= bitmask;
+	}
+
+	sys_write32(reg_val, config->base_address + rmstpsr[reg]);
+	if (!enable) {
+		cpg_reset(config, reg, bit);
+	}
+
+	irq_unlock(key);
+
+	return 0;
+}
+
+static int cpg_mssr_blocking_start(const struct device *dev,
+			       clock_control_subsys_t sys)
+{
+	struct rcar_cpg_clk *clk = (struct rcar_cpg_clk *) sys;
+	int ret = -EINVAL;
+
+	if (clk->domain == CPG_MOD) {
+		ret = cpg_rmstp_clock_endisable(dev, clk->module, true);
+	} else if (clk->domain == CPG_CORE) {
+		ret = cpg_core_clock_endisable(dev, clk->module, clk->rate,
+					       true);
+	}
+
+	return ret;
+}
+
+static int cpg_mssr_stop(const struct device *dev,
+		     clock_control_subsys_t sys)
+{
+	struct rcar_cpg_clk *clk = (struct rcar_cpg_clk *) sys;
+	int ret = -EINVAL;
+
+	if (clk->domain == CPG_MOD) {
+		ret = cpg_rmstp_clock_endisable(dev, clk->module, false);
+	} else if (clk->domain == CPG_CORE) {
+		ret = cpg_core_clock_endisable(dev, clk->module, 0, false);
+	}
+
+	return ret;
+}
+
+static int cpg_get_rate(const struct device *dev,
+			clock_control_subsys_t sys,
+			uint32_t *rate)
+{
+	const struct rcar_mssr_config *config = DEV_CFG(dev);
+	struct rcar_cpg_clk *clk = (struct rcar_cpg_clk *) sys;
+	uint32_t val;
+	int ret = 0;
+
+	if (clk->domain != CPG_CORE) {
+		return -ENOTSUP;
+	}
+
+	switch (clk->module) {
+	case CPG_CORE_CLK_CANFD:
+		val = sys_read32(config->base_address + CANFDCKCR);
+		if (val & CANFDCKCR_CKSTP) {
+			*rate = 0;
+		} else {
+			val &= CANFDCKCR_DIVIDER_MASK;
+			*rate = CANFDCKCR_PARENT_CLK_RATE / (val + 1);
+		}
+		break;
+	case CPG_CORE_CLK_S3D4:
+		*rate = S3D4_CLK_RATE;
+		break;
+	default:
+		ret = -ENOTSUP;
+	}
```

Comment:
Although not required, `break` here would be nice since this is not an intentional fall-through.

p.s. are we not getting compiler warnings about this?

Reply:
at least with arm-zephyr-eabi-gcc (crosstool-NG 1.24.0.37-3f461da-dirty) 9.2.0 I don't have any warning.
will add break statement